ceph: 9.2.0 -> 12.2.2

adevress opened the pull request:
- Update Ceph to last version
- Solve security vulnerability of the old version
- Provide shared version of snappy, required for Ceph

###### Motivation for this change
- Make Ceph usable again
- Provide shared version for snappy

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

andir commented:
Thank you for working on this. You must also update the sha256 of ceph otherwise you will build the old version with a new name ;)

Mic92 commented on pkgs/development/libraries/snappy/default.nix:
```diff
@@ -1,4 +1,4 @@
-{ stdenv, fetchFromGitHub, cmake }:
+{ stdenv, fetchFromGitHub, cmake, sharedLibrary ? false }:
```
I would rather build shared libraries by default. Which is what we prefer in nixpkgs.

Mic92 replied:
This variable does not have an effect on this derivation, was this added by mistake?

adevress commented:
Hi @Mic92,

Ok this patch comes from an old nixpkgs branch we have in our side (BBP).

I just realized it diverged way more I expected, and that the recommended way to build Ceph is cmake nowadays.

This would imply way more modifications, let me have a look to that and I come back to you.

adevress committed:
ceph: 9.2.0 -> 12.2.2

- Update Ceph to last version
- Solve vulernatibility problem in the old version
- Patch rocksdb ceph rocksdb API
- Simplify a lot ceph build by switching to CMake

This Pull request require #34662 and #34661 to be effective

adevress commented:
ping @vcunat  @Mic92 

Ok guys, problems fixed.


- The build reciep for Ceph has been entirely switched to CMake. It greatly simplified the build process.
- External dependencies to boost and RocksDB are enforced.
- This pull request requires #34662 and #34661 to be merged
- Ceph is now built with RocksDB as default backend
- Ceph now has its own python environment, avoiding ugly hack to get its python tool / doc generation working.
- Every output is separated correctly

vcunat commented:
/cc maintainer @alexanderkjeldaas.  BTW, it's good to keep `lib.maintainers.*` aligned with github user names.

srhb commented:
@lejonet Might be interesting to test this as the base for your ceph module :)h

lejonet commented:
@srhb Yeah, I will do that, seems like @adevress did a thorough work through of the build, I only fixed what I needed to get it to compile at all, to develop the ceph module.

Will fork and apply my module ontop and try it out

lejonet commented:
@adevress I've tried your branch on a nixos system, with sandbox enabled and all binaries compile and run fine
I just had to make a tiny patch for the ceph source code and your package that you can see here: https://github.com/lejonet/nixpkgs/commit/a782a50ad160bc184e5ba972a1eb8cf77746bcce

joachifm commented:
@GrahamcOfBorg build ceph bareos

GrahamcOfBorg commented:
Failure on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.33894&attempt_id=716dc9b5-a953-4ed2-b3fd-760fc40380a3)

<details><summary>Partial log (click to expand)</summary><p>

```
Package ‘lttng-ust-2.10.0’ in /private/var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/lnl7-mac/pkgs/development/tools/misc/lttng-ust/default.nix:32 is not supported on ‘x86_64-darwin’, refusing to evaluate.

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowBroken = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowBroken = true; }
to ~/.config/nixpkgs/config.nix.

```
</p></details>

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.33894&attempt_id=29657cf0-0fdb-4f50-8188-a80c0477ec4d)

<details><summary>Partial log (click to expand)</summary><p>

```
remote: Total 257 (delta 43), reused 82 (delta 10), pack-reused 0
Receiving objects: 100% (257/257), 697.92 KiB | 11.26 MiB/s, done.
Resolving deltas: 100% (43/43), done.
From https://github.com/facebook/zstd
 * branch            refs/pull/484/head -> FETCH_HEAD
Switched to a new branch 'fetchgit'
removing `.git'...
cannot build derivation '/nix/store/7a9c0pkgx7sxrfnbjlpkrw0q28ry8d9z-ceph-12.2.2.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/4iyilmysdm0wjy7p1d54bzflg9jql709-bareos-15.2.4.drv': 2 dependencies couldn't be built
error: build of '/nix/store/4iyilmysdm0wjy7p1d54bzflg9jql709-bareos-15.2.4.drv', '/nix/store/7a9c0pkgx7sxrfnbjlpkrw0q28ry8d9z-ceph-12.2.2.drv' failed
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.33894&attempt_id=85a9ce4b-f06b-4400-9a36-fdb80cfe3335)

<details><summary>Partial log (click to expand)</summary><p>

```
[ 25%] Building CXX object src/mds/CMakeFiles/mds.dir/Capability.cc.o
[ 25%] Building CXX object src/mds/CMakeFiles/mds.dir/MDSDaemon.cc.o
[ 25%] Building CXX object src/mds/CMakeFiles/mds.dir/MDSRank.cc.o
[ 25%] Building CXX object src/mds/CMakeFiles/mds.dir/Beacon.cc.o
[ 25%] Building CXX object src/mds/CMakeFiles/mds.dir/flock.cc.o
[ 25%] Building C object src/mds/CMakeFiles/mds.dir/locks.c.o
[ 25%] Building CXX object src/mds/CMakeFiles/mds.dir/journal.cc.o
building of ‘/nix/store/hyjyagxn8y7wwjgk20h1hsx4agh1v00b-ceph-12.2.2.drv’ timed out after 3600 seconds
cannot build derivation ‘/nix/store/78a2lv0pqwkx1gk2bz5z5ypak4jn50lv-bareos-15.2.4.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/78a2lv0pqwkx1gk2bz5z5ypak4jn50lv-bareos-15.2.4.drv’, ‘/nix/store/hyjyagxn8y7wwjgk20h1hsx4agh1v00b-ceph-12.2.2.drv’ failed
```
</p></details>

Mic92 commented:
Thanks!
